---
permalink: /
title: ""
excerpt: "About me"
author_profile: true
redirect_from: 
  - /about/
  - /about.html
---

Hi! I'm Aqib (pronounced ah-kib), a mechanical engineer turned computational and synthetic biologist. Currently, I am a graduate student in the [Biological Control Group](https://yeung.me.ucsb.edu/). My research takes inspiration from control and dynamical systems to develop computational tools for studying and designing synthetic biological circuits and sensors. I have developed numerous synthetic biological devices to validate developed computational and theoretical methods. My current interests are in developing matrix factorization techniques to integrate time-series, multi-modal 'omic data with application in i) understanding how and why synthetic gene circuits impart mechanisms of adaptation in microbes, and ii) how single-gene mutant co-fitnesses can be mapped to optimal multi-gene mutant fitness. I have experience in many facets of machine learning and have made contributions to a relatively new dimensionality reduction technique known as dynamic [mode decomposition](https://en.wikipedia.org/wiki/Dynamic_mode_decomposition). I am also interested and am proficient in the developing algorithms for analysis of single-cell RNA-seq and multi-omics. I developed single-cell dynamic mode decomposition (at my time at Cellarity) for feature extraction, improving the mapping from single-cell dynamics to phenotypes in the context of disease biology. 

Feel free to reach out! 

![dmd](/images/dmd_proj_yeast.png)  
*My graduate research largely focused on learning dynamics from large biological datasets at the intersection of machine learning and systems biology. This image depicts the transcriptome dynamics of a yeast mutant, lacking key cell cycle genes, projected in two dimensions by dynamic mode decomposition. The modes define a reduced dimension space in which, for the two coordinates shown here, separate genes which respond to the mutations and those which do not.* 
